File: 3-LongestSubstring/longestSubstring.cs
// https://leetcode.com/problems/longest-substring-without-repeating-characters/
// https://www.youtube.com/watch?v=fTIzYJvhsqg

// Its a sliding window problem! you did this at LI-COR

public class Solution
{
    public int LengthOfLongestSubstring(string s)
    {
        int start = 0;
        int end = 0;
        int maxLen = 0;
        HashSet<char> charSet = new();

        while (end < s.Length)
        {
            if (!charSet.Contains(s[end]))
            {
                charSet.Add(s[end]);
                end++;
                maxLen = Math.Max(maxLen, end - start);
            }
            else
            {
                charSet.Remove(s[start]);
                start++;
            }
        }
        return maxLen;
    }
}

File: 3-LongestSubstring/longestSubstring.py

def lengthOfLongestSubstring(s: str) -> int:
    start: int = 0
    end: int = 0
    max_len: int = 0
    char_set: list = []

    while end < len(s):
        if end in char_set:
            char_set.remove(s[start])
            start += 1
        else:
            char_set.append([])
            end += 1
            max_len = max(max_len, end - start)

    return max_len
